### ISSUE TITLE
AWGN noise_variance dependence on FEC code rate

### ISSUE THREAD

**Issue by aklsh** [veeresht/CommPy#66]
Hey,
This isn't an issue with the code, but something I want to understand the theory behind, and I couldn't find a better place to ask.

https://github.com/veeresht/CommPy/blob/1d26b074818ad3a56fe7baa70c92b04ab575f612/commpy/channels.py#L699

In an AWGN channel, why does the noise variance depend on the FEC code rate? If you can share some references as to where you found this, it will be very helpful. (I am trying to do a self-study course on Coding theory where I am trying to code and implement the concepts I learn... and I'm trying to implement different channel models to understand better.)

**Comment by rafaelmcam** [comment]
@aklsh 

The noise variance "depends" on the code rate in the sense that we cannot compare two transmissions with two different code rates directly. The traditional performance measure is the bit error rate (BER) vs signal to noise ratio (Eb/N0), the key point is in what we usually mean by "signal to noise ratio", in this case it's based on INFORMATION bits and not transmitted bits. So if we use a code 4/7 code we will transmit 7 bits per block but in fact only 4 bits are information bit.

An intuitive explanation of that is as follows: Imagine we have an uncoded transmission where we desire to "use" 1 "Joule" (an abstract energy unit actually) per transmitted bit. The total transmission length is 400 bits so we used 400J and achieved a BER = x. Now if we compare with a coded transmission the same data will be encoded by a 4/7 rate code, so we now have 700 bits to be transmitted, we have several option on how to do this transmission, let's assume we will use 1J energy per TRANSMITTED bit (this assumes the digital channel is the "same" as the uncoded transmission, so same noise variance per simulated transmitted bit), in this case we just used 700J of energy for the entire packet in opposition to 400J in the uncoded case. This clearly indicates that we were simply doing a unfair comparison. If we use the same "digital channel" we are actually implicitly biasing the performance measure in favor of the coded transmission. We need to spend the same amount of energy on both packet transmission for a fair result in the discrete case otherwise we could simply have transmitted the uncoded version with more energy per bit which would result in a smaller BER. Increasing the noise variance for rate != 1 codes by the exposed relation corrects this bias so we can compare BER performances with codes with different rates.

In a related note, error coding archives "real coding gain" in the sense that the performance improvement coming out of using error correcting codes are not SOLELY based on the fact that we are transmitting more bits for better performance, but actually the codes archives gains PROPORTIONALLY to the information bits! Which is the main point about using error coding.

A useful reference about this is

(Pages 15~20 - Section 1.7) S. Lin and D. J. Costello, Error Control Coding, 2nd ed. Upper Saddle River, N.J: Prentice Hall, 2004.
